test/python/transpiler/test_swap_strategy.py

```diff
@@ -0,0 +1,256 @@
+            SwapStrategy(
+                coupling_map=self.line_coupling_map,
+                swap_layers=[[(0, 1), (1, 2)], [(1, 3), (2, 4)]],
+            )
+
+    def test_distance_matrix(self):
+        """Test the computation of the swap strategy distance matrix."""
+        line_distance_matrix = [
+            [0, 0, 3, 1, 2],
+            [0, 0, 0, 2, 3],
+            [3, 0, 0, 0, 1],
+            [1, 2, 0, 0, 0],
+            [2, 3, 1, 0, 0],
+        ]
+        self.assertEqual(line_distance_matrix, self.line_strategy.distance_matrix)
+
+    def test_reaches_full_connectivity(self):
+        """Test to reach full connectivity on the longest line of Mumbai."""
+
+        # The longest line on e.g. Mumbai has the qubits
+        ll27 = [
+            0,
+            1,
+            2,
+            3,
+            5,
+            8,
+            11,
+            14,
+            16,
+            19,
+            22,
+            25,
+            24,
+            23,
+            21,
+            18,
+            15,
+            12,
+            10,
+            7,
+            6,
+        ]
+
+        ll27_map = [[ll27[idx], ll27[idx + 1]] for idx in range(len(ll27) - 1)]
+        ll27_map += [[ll27[idx + 1], ll27[idx]] for idx in range(len(ll27) - 1)]
+
+        # Create a line swap strategy on this line
+        layer1 = [(ll27[idx], ll27[idx + 1]) for idx in range(0, len(ll27) - 1, 2)]
+        layer2 = [(ll27[idx], ll27[idx + 1]) for idx in range(1, len(ll27), 2)]
+
+        n = len(ll27)
+        for n_layers, result in [
+            (n - 4, False),
+            (n - 3, False),
+            (n - 2, True),
+            (n - 1, True),
+        ]:
+            swap_strat_ll = []
+            for idx in range(n_layers):
+                if idx % 2 == 0:
+                    swap_strat_ll.append(layer1)
+                else:
+                    swap_strat_ll.append(layer2)
+
+            strat = SwapStrategy(CouplingMap(ll27_map), swap_strat_ll)
+            self.assertEqual(strat.reaches_full_connectivity, result)
+
+    def test_possible_edges(self):
+        """Test that possible edges works as expected."""
+        coupling_map = CouplingMap(couplinglist=[(0, 1), (1, 2), (2, 3)])
+
+        strat = SwapStrategy(coupling_map, [[(0, 1), (2, 3)], [(1, 2)]])
+
+        expected = set()
+        for i in range(4):
+            for j in range(4):
+                if i != j:
+                    expected.add((i, j))
+
+        self.assertSetEqual(strat.possible_edges, expected)
+
+
+class TestSwapStrategyExceptions(QiskitTestCase):
+    """A class to test the exceptions raised by swap strategies."""
+
+    def test_invalid_strategy(self):
+        """Test that a raise properly occurs."""
+
+        coupling_map = CouplingMap(couplinglist=[(0, 1), (1, 2)])
+        swap_layers = [[(0, 1), (2, 3)], [(1, 2), (3, 4)]]
+
+        with self.assertRaises(QiskitError):
+            SwapStrategy(coupling_map, swap_layers)
```

Comment:
Here (and in other tests), it can be good to use `self.assertRaisesRegex` to ensure that the right message is being shown to users.

(But thanks for putting in the specific tests of the failure paths - most PRs don't do that and they're super important!)

Reply:
Made a few changes here: [bf22a2d](https://github.com/Qiskit/qiskit-terra/pull/7813/commits/bf22a2d0a7c5bafdf8561fc37d05fa4a289efd7f)